File: weasel.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
 
const char target[] = "METHINKS IT IS LIKE A WEASEL";
const char tbl[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ ";
 
#define CHOICE (sizeof(tbl) - 1)
#define MUTATE 15
#define COPIES 50
 
/* retorna um inteiro aleatório de 0 a n - 1 */
int irand(int n)
{
	int r, rand_max = RAND_MAX - (RAND_MAX % n);
	while((r = rand()) >= rand_max);
	return r / (rand_max / n);
}
 
/* número de caracteres diferentes entre a e b */
int unfitness(const char *ancestral, const char *prole)
{
	int i, sum = 0;
	for (i = 0; ancestral[i]; i++)
		sum += (ancestral[i] != prole[i]);
	return sum;
}
 
/* cada caractere de b tem 1 / MUTATE de chance de diferir de a */
void mutate(const char *ancestral, char *prole)
{
	int i;
	for (i = 0; ancestral[i]; i++)
		prole[i] = irand(MUTATE) ? ancestral[i] : tbl[irand(CHOICE)];
 
	prole[i] = '\0';
}
 
int main (int argc, char** argv)

{
	int i, best_i, unfit, best, iters = 0;
	char specimen[COPIES][sizeof(target) / sizeof(char)];
 
	/* sorteando a string inicial */
	for (i = 0; target[i]; i++)
		specimen[0][i] = tbl[irand(CHOICE)];
	specimen[0][i] = 0;
 
	do {
		for (i = 1; i < COPIES; i++)
			mutate(specimen[0], specimen[i]);
 
		/* encontrando a string mais parecida com o alvo(target) */
		for (best_i = i = 0; i < COPIES; i++) {
			unfit = unfitness(target, specimen[i]);
			if(unfit < best || !i) {
				best = unfit;
				best_i = i;
			}
		}
 
		if (best_i) strcpy(specimen[0], specimen[best_i]);
		printf("iter %d, score %d: %s\n", iters++, best, specimen[0]);

	} while (best);
 
	return 0;
}

File: weasel.py
# Implementação em Python do programa weasel de Richard Dawkin
# Ilustra o poder da seleção cumulativa em comparação com a pesquisa inteiramente aleatória

# Inicializando
import random
target = list("METHINKS IT IS LIKE A WEASEL")   # Sequência alvo
alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ "        # Alfabeto de símbolos possíveis (letras maiúsculas + espaço em branco)
n_offspring = 50                                # Número de descendentes por geração
mut_rate = 0.09                                 # Taxa de mutação: probabilidade de que qualquer letra mude
best_offspring = []                             # Variável contendo a melhor prole após cada geração


# Gerando a string inicial
parent = []
for i in range(len(target)):
    parent.append(random.choice(alphabet))

# Main loop: construir descendentes mutantes, selecionar o melhor para a próxima geração, parar quando o alvo for atingido
gen = 0
while best_offspring != target:
    gen = gen + 1

    # Gerar n_offspring proles que podem conter mutacoes
    kid_list = []
    for i in range(n_offspring):

        # Copiar conteúdo de pai para filho
        kid = parent[:]

        # Repita as posições na prole, pois cada posição permite a possibilidade de mutação
        kid_changed = False
        for pos in range(len(kid)):

            # Deixe o resíduo sofrer mutação com probabilidade mut_rate
            if random.random() < mut_rate:
                kid_changed = True

                # Selecione aleatoriamente um novo símbolo (que é diferente do que já está presente)
                old_symbol = parent[pos]
                possible_new_symbols = set(alphabet) - set(old_symbol)
                new_symbol = random.choice(list(possible_new_symbols))

                # Mutacao prole
                kid[pos] = new_symbol

                
        # Adicionar (possivelmente) prole com mutação à lista de proles atuais
        kid_list.append(kid)


    # Encontre a prole mais adequada (= string mais semelhante ao alvo)
    smallest_dif = len(target) + 1
    for kid in kid_list:

        # Encontre o número de posições que diferem entre a prole e o alvo

        dif = 0.0
        for pos in range(len(target)):
            if kid[pos] != target[pos]:
                dif = dif + 1

        # Mantenha a melhor prole encontrada até agora
        if dif < smallest_dif:
            smallest_dif = dif
            best_offspring = kid

    # considere a melhor prole como ponto de partida para a próxima rodada de mutação

    parent = best_offspring

    # Impressão do progresso
    fitness = (len(target)-smallest_dif)/len(target)            # Aptidão do indivíduo mais parecido com o alvo
    result_string = ""
    for pos in range(len(target)):
        if best_offspring[pos] == target[pos]:
            result_string += best_offspring[pos]
        else:
            result_string += best_offspring[pos].lower()
    print ("%s     ** Gen: %4d   Diferentes: %3d   Aptidao: %.4f  " % (result_string,gen, smallest_dif, fitness))

